plenum/server/future_primaries_batch_handler.py

```diff
@@ -0,0 +1,99 @@
+import copy
+from collections import OrderedDict
+from typing import Tuple, Dict
+
+from common.exceptions import LogicError
+from plenum.common.constants import TXN_TYPE, NODE, TARGET_NYM, ALIAS, SERVICES, VALIDATOR, DATA, POOL_LEDGER_ID
+from plenum.common.util import getMaxFailures
+from plenum.server.batch_handlers.batch_request_handler import BatchRequestHandler
+from plenum.server.batch_handlers.three_pc_batch import ThreePcBatch
+
+
+class NodeState:
+    def __init__(self, node_reg, node_ids, primaries):
+        self.node_reg = node_reg
+
+        # Nodes are never delete from here, so they are in a sequential order
+        self.node_ids = node_ids
+
+        # List of future primaries
+        self.primaries = primaries
+
+        self.number_of_inst = FuturePrimariesBatchHandler. \
+            get_required_number_of_instances(len(node_reg))
+
+
+class FuturePrimariesBatchHandler(BatchRequestHandler):
+    # This class is needed for correct primaries storing in audit ledger.
+    # It is something like uncommitted state of primaries.
+    # It emulate primaries changes when pool txn is just applied.
+    # When current batches reverts, it drops it's properties to node's.
+
+    def __init__(self, database_manager, node):
+        super().__init__(database_manager, POOL_LEDGER_ID)
+        self.node = node
+        self.node_states = []
+
+    def post_batch_applied(self, three_pc_batch: ThreePcBatch, prev_handler_result=None):
+        node_txn_count = 0
+        last_state = None
+        if len(self.node_states) == 0:
+            last_state = self.create_node_state_from_current_node()
+        else:
+            last_state = copy.deepcopy(self.node_states[-1])
```

Comment:
It seems like this could be imlemented in a much more safe (and probably efficient) way if `NodeState` was immutable `NamedTuple` instead of mutable class with fields.